Fix endpoint slice filtering to ensure we get all the necessary objects

odinuge opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
Fix endpoint slices filtering to ensure we filter out headless services and continue to support older k8s versions where service labels are not propagated to endpoint slices
```

odinuge commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -567,7 +572,7 @@ func (k *K8sWatcher) enableK8sWatchers(ctx context.Context, resourceNames []stri
 			// no-op; handled in resources.K8sAPIGroupEndpointV1Core.
 		case resources.K8sAPIGroupEndpointV1Core:
-			k.initEndpointsOrSlices(k.clientset.Slim(), serviceOptModifier)
+			k.initEndpointsOrSlices(k.clientset.Slim(), serviceAndEndpointOptModifier, endpointSliceOptModifier)
```
Not sure if we should plumb it in here or just just run `GetEndpointSliceListOptionsModifier` where we need it.

joamaki replied:
Fine here. The PR https://github.com/cilium/cilium/pull/23977 removes this from here anyway. I'll rebase and incorporate these changes once this PR lands.

odinuge-work committed:
Fix endpoint slice filtering when talking to Kubernetes

This fixes the filtering of endpoint slices to ensure that we support
all the k8s versions we intend to. This ensures that we always filter
out endpoint slices with the well-known "headless" label, and
_do not_ filter out any endpoint slices based on the service proxy
label.

In pre Kubernetes v1.20, the labels on a service were not mirrored into
the labels of the endpoint slice. The headless label was not applied.
See PR 94443 in kubernetes/kubernetes for more info.

When no longer supporting Kubernetes v1.20, we can remove this custom
logic - and use the same label filter for endpoints, services and endpoint
slices.

Historically, we had no filters on the endpoint slice objects, but with
the two referred commits, the same filter we had for endpoints and
services was introduced to endpoint slices as part of the refactor. The
reason we don't revert the behavior directly, is that we _do want_ to filter
out endpoint slices for headless services, like we do with normal endpoints.

For completeness; the end user behavior will now be equal for both
endpoints and endpoint slices; since we will always filter the services
in the same way, and when we get an endpoint slice without a
corresponding service in state, we effectively ignore that endpoint slice.

Fixes: ca3a4df01023 ("k8s: Add Resource[*Endpoints] to shared resources")
Fixes: 82a728aa1780 ("agent, operator, clustermesh-apiserver: use Resource[*Endpoints]")
Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>
